Unexpected type error because missing capture conversion

Take:

```java
class IncompatibleTypes {
    void possibleValues(final Class<? extends Enum> enumType) {
        lowercase(enumType.getEnumConstants());
    }

    <T extends Enum<T>> void lowercase(final T... items) {
    }
}
```

Running any checker on this file will result in:

```
error: [argument.type.incompatible] incompatible types in argument.
        lowercase(enumType.getEnumConstants());
                                           ^
  found   : ?[ extends @UnknownKeyFor Enum super @KeyForBottom Void] @UnknownKeyFor []
  required: ?[ extends @UnknownKeyFor Enum super @KeyForBottom Void] @UnknownKeyFor []
```

The wildcard should be captured properly and no error is expected here.

(The following three checkers produce different errors which should also be investigated:

```
GUI Effect Checker (no imcompatible error, passed)
Lock Checker
	found : ? extends @GuardedBy @LockPossiblyHeld Enum @GuardedBy @LockPossiblyHeld []
	 required: @LockPossiblyHeld ? extends @GuardedBy @LockPossiblyHeld Enum @GuardedBy @LockPossiblyHeld []
Aliasing Checker
	found : ? extends @MaybeAliased @MaybeLeaked Enum @MaybeAliased @MaybeLeaked []
	required: @MaybeLeaked ? extends @MaybeAliased @MaybeLeaked Enum @MaybeAliased @MaybeLeaked []
```
)

Thanks for reporting this issue.  Your test case issues javac warnings:
````
javacheck -processor interning Issue759.java -Xlint:unchecked
Issue759.java:10: warning: [unchecked] unchecked method invocation: method lowercase in class IncompatibleTypes is applied to given types
        lowercase(enumType.getEnumConstants());
                 ^
  required: T[]
  found: CAP#1[]
  where T is a type-variable:
    T extends Enum<T> declared in method <T>lowercase(T...)
  where CAP#1 is a fresh type-variable:
    CAP#1 extends Enum from capture of ? extends Enum
````
Generally if javac issues an unchecked warning, the Checker Framework is likely to issue a type-checking error at the same location. However, the test case below, which is similar to your test case, also issues an unexpected type-checker error.  I've added the test case to the repository.
````
class IncompatibleTypes {
    void possibleValues(final Class<? extends Enum> enumType) {
        lowercase(enumType.getEnumConstants());
    }
    <T> void lowercase(final T items) {
    }
}
````

--- a/checker/tests/nullness/Issue759.java
+++ b/checker/tests/nullness/Issue759.java
@@ -1,0 +1,54 @@
+// Testcase for Issue759
+// https://github.com/typetools/checker-framework/issues/759
+// Also, see framework/tests/all-systems/Issue759.java
+import org.checkerframework.checker.nullness.qual.*;
+@SuppressWarnings("unchecked")
+public class Issue759 {
+    void possibleValues(final Class<? extends Enum> enumType) {
+        lowercase(enumType.getEnumConstants());
+        lowercase2(enumType.getEnumConstants());
+        lowercase3(enumType.getEnumConstants());
+    }
+
+    <T extends Enum<T>> void lowercase(final T @Nullable ... items) {
+    }
+
+    <T extends Enum<T>> void lowercase2(final T @Nullable [] items) {
+    }
+
+    <T> void lowercase3(final T items) {
+    }
+}
+
+interface Gen<T extends Gen<T>> {
+    T[] getConstants();
+    T @Nullable [] getNullableConstants();
+}
+
+class IncompatibleTypes {
+    void possibleValues(final Gen<?> genType) {
+        lowercase(genType.getConstants());
+        lowercase(genType.getNullableConstants());
+    }
+
+    <S> void lowercase(final S items) {
+    }
+
+    void possibleValues2(final Gen<?> genType) {
+        lowercase2(genType.getConstants());
+        //:: error: (type.argument.type.incompatible)
+        lowercase2(genType.getNullableConstants());
+    }
+
+    <S extends Object> void lowercase2(final S items) {
+    }
+
+    void possibleValues3(final Gen<?> genType) {
+        lowercase3(genType.getConstants());
+        //:: error: (argument.type.incompatible)
+        lowercase3(genType.getNullableConstants());
+    }
+
+    <S> void lowercase3(final @NonNull S items) {
+    }
+}

--- a/framework/src/org/checkerframework/framework/type/DefaultTypeHierarchy.java
+++ b/framework/src/org/checkerframework/framework/type/DefaultTypeHierarchy.java
@@ -29,6 +29,7 @@ import java.util.Set;
 import javax.lang.model.element.AnnotationMirror;
 import javax.lang.model.element.TypeElement;
 import javax.lang.model.type.TypeKind;
+import javax.lang.model.type.TypeMirror;
 import javax.lang.model.util.Elements;
 import javax.lang.model.util.Types;
 
@@ -779,6 +780,27 @@ public class DefaultTypeHierarchy extends AbstractAtmComboVisitor<Boolean, Visit
 
     @Override
     public Boolean visitWildcard_Wildcard(AnnotatedWildcardType subtype, AnnotatedWildcardType supertype, VisitHistory visited) {
+        // This method is called at a method invocation where a type variable in the method
+        // declaration is substituted with a wildcard.
+        // For example:
+        // <T> void method(Gen<T> t) {}
+        // Gen<?> x;
+        // method(x); // this method is called when checking this method call
+
+        if (subtype.getUnderlyingType() == supertype.getUnderlyingType()) {
+            boolean subtypeHasAnno = subtype.getAnnotationInHierarchy(currentTop) != null;
+            boolean supertypeHasAnno = supertype.getAnnotationInHierarchy(currentTop) != null;
+
+            if (subtypeHasAnno && supertypeHasAnno) {
+                // if both have primary annotations then just check the primary annotations
+                // as the bounds are the same
+                return isPrimarySubtype(subtype, supertype, true);
+
+            } else if (!subtypeHasAnno && !supertypeHasAnno && areEqualInHierarchy(subtype, supertype, currentTop)) {
+                // two unannotated uses of the same wildcard are of the same type
+                return true;
+            }
+        }
         return visitWildcardSubtype(subtype, supertype, visited);
     }
 

--- a/framework/tests/all-systems/Issue759.java
+++ b/framework/tests/all-systems/Issue759.java
@@ -1,16 +1,13 @@
 // Testcase for Issue759
 // https://github.com/typetools/checker-framework/issues/759
-// @skip-test
+@SuppressWarnings({"nullness", "unchecked"}) // See checker/test/nullness/Issue759.java
 public class Issue759 {
     void possibleValues(final Class<? extends Enum> enumType) {
-        //:: warning: [unchecked] unchecked method invocation: method lowercase in class Issue759 is applied to given types
         lowercase(enumType.getEnumConstants());
-        //:: warning: [unchecked] unchecked method invocation: method lowercase2 in class Issue759 is applied to given types
         lowercase2(enumType.getEnumConstants());
         lowercase3(enumType.getEnumConstants());
     }
 
-    //:: warning: [unchecked] Possible heap pollution from parameterized vararg type T
     <T extends Enum<T>> void lowercase(final T... items) {
     }
 
@@ -20,11 +17,12 @@ public class Issue759 {
     <T> void lowercase3(final T items) {
     }
 }
-
+@SuppressWarnings("nullness")
 class Gen<T extends Gen<T>> {
     T[] getConstants() { return null; }
 }
 
+@SuppressWarnings("nullness")
 class IncompatibleTypes {
     void possibleValues(final Gen<?> genType) {
         lowercase(genType.getConstants());

